tlatools/org.lamport.tlatools/src/tlc2/tool/ITool.java

```diff
@@ -127,6 +129,17 @@ public interface ITool extends TraceApp, OpDefEvaluator {
 	TLCState enabled(SemanticNode pred, IActionItemList acts, Context c, TLCState s0, TLCState s1);
 	TLCState enabled(SemanticNode pred, IActionItemList acts, Context c, TLCState s0, TLCState s1, CostModel cm);
 
+	/**
+	 * Evaluate the given invariant in the given state and, if it evaluates to FALSE, return information about why.
+	 *
+	 * @param invariant the invariant
+	 * @param s0        the current state
+	 * @param s1        the next state (or {@link TLCState#Empty})
+	 * @return information about why the invariant evaluates to FALSE, or {@code null} if the invariant evaluates to TRUE
+	 * @throws util.Assert.TLCRuntimeException if the invariant does not evaluate to a boolean value
+	 */
+	InvariantViolation findFalsifyingContext(Action invariant, TLCState s0, TLCState s1);
```

Comment:
Could this be generalized to `ITool#findInvalidAssignment*` methods to align with the existing `ITool#isValid*` methods? I would even be supportive of a refactoring of the return values of the various `isValid` methods. Additionally, should we introduce a constant level variant that accepts an `ExprNode` instead of an `Action`?